File: 2022/day/2/Main.hs
import Control.Arrow
import Data.Char
import Data.Monoid


data Shape
    = Rock
    | Paper
    | Scissor
    deriving (Eq, Ord, Show)


shapeScore Rock = 1
shapeScore Paper = 2
shapeScore Scissor = 3

data Round = Round Shape Shape

decode :: Char -> Shape
decode 'A' = Rock
decode 'B' = Paper
decode 'C' = Scissor
decode 'X' = Rock
decode 'Y' = Paper
decode 'Z' = Scissor

win Rock = Paper
win Paper = Scissor
win Scissor = Rock

lose Rock = Scissor
lose Paper = Rock
lose Scissor = Paper

strategy1 = score

strategy2 (Round p1 p2) =
    case p2 of
        Rock -> score (Round p1 (lose p1))
        Paper -> score (Round p1 p1)
        Scissor -> score (Round p1 (win p1))

score (Round p1 p2) = shapeScore p2 + outcomeScore
  where
    outcomeScore
      | win p1 == p2 = 6
      | p1 == p2 = 3
      | otherwise = 0

parseRound :: MonadFail m => String -> m Round
parseRound [p, ' ', q] = pure (Round (decode p) (decode q))
parseRound _           = fail "bad input"

main = do
    input <- readFile "input.txt"
    rounds <- traverse parseRound (lines input)

    -- Part 1
    print $ getSum $ foldMap (Sum . strategy1) rounds

    -- Part 2
    print $ getSum $ foldMap (Sum . strategy2) rounds

File: 2022/day/2/main.py
ans1 = 0
ans2 = 0

with open("input.txt") as fp:
    for ln in fp:
        p1, p2 = ln[0], ln[2]
        
        # Convert to Rock=0, Paper=1, Scissor=2
        q1, q2 = ord(p1) - ord('A'), ord(p2) - ord('X')

        if (q1 + 1) % 3 == q2:  # Win
            ans1 += 6 + q2 + 1
        elif q1 == q2:  # Draw
            ans1 += 3 + q2 + 1
        else:  # Lose
            ans1 += 0 + q2 + 1

        if q2 == 0:  # Lose
            ans2 += 0 + (q1 - 1) % 3 + 1
        elif q2 == 1:  # Draw
            ans2 += 3 + q1 + 1
        else:  # Win
            ans2 += 6 + (q1 + 1) % 3 + 1

print(ans1)
print(ans2)
